### DISCUSSION TITLE
ATR Trailing Stops

### DISCUSSION THREAD

**Discussion by myalgomate** [DaveSkender/Stock.Indicators#724] [Indicators]
Community discussion for [ATR Trailing Stop](https://dotnet.stockindicators.dev/indicators/AtrStop/#content)

**Comment by myalgomate** [comment]
Can you please check for the indicator ATR Trailing Stops
Here I am sharing u some links to evaluate more
https://www.stockmaniacs.net/atr-trailing-stops/

https://library.tradingtechnologies.com/trade/chrt-ti-atr-trailing-stops.html#:~:text=The%20ATR%20Trailing%20Stops%20indicator,narrower%20stops%20indicate%20less%20volatility.

[ATR-Stop-Loss-Calculator.xlsx](https://github.com/DaveSkender/Stock.Indicators/files/8167172/ATR-Stop-Loss-Calculator.xlsx)
Sharing a Calculation of TCS 15 min OHLC data with Calculation for your easy reference.

**Comment by DaveSkender** [comment]
This is very similar to [SuperTrend](https://daveskender.github.io/Stock.Indicators/indicators/SuperTrend/#content) or maybe [Volatility Stop](https://daveskender.github.io/Stock.Indicators/indicators/VolatilityStop/#content).

**Reply by DaveSkender** [reply]
In doing some research, I see this [ATR Trailing Stop](https://www.stockopedia.com/learn/charts-technical-analysis/atr-trailing-stops-462698/) is originally authored by Welles Wilder, so I should be able to find the original source recipe from his publications.

I think this pre-dates Oliver Seban’s SuperTrend and uses close price where SuperTrend used HL2, but is otherwise identical.

**Reply by DaveSkender** [reply]
I’m adding to the backlog.  See #729 for tracking.

**Reply by DaveSkender** [reply]
In looking at Wilder’s book _New Concepts in Technical Trading Systems_ I think this is very similar to - [Volatility Stop](https://daveskender.github.io/Stock.Indicators/indicators/VolatilityStop/#content), [Parabolic SAR](https://daveskender.github.io/Stock.Indicators/indicators/ParabolicSar/#content), and [SuperTrend](https://daveskender.github.io/Stock.Indicators/indicators/SuperTrend/#content) (already in the library).  ATR Trailing Stop would likely be a more parameterized variant that allows options to select `CandlePart` and `EndType` to accommodate user preferences.

**Reply by DaveSkender** [reply]
On implementing, I determined that `CandlePart` cannot be a chosen parameter as this is always triggered by Close price; it also gets complicated when allowing `EndType` arguments as it sets bandwidths based on either Close or High/Low.  Given that ATR requires the whole candle range, it would be a bad idea to customize in that manner, IMO.

**Comment by myalgomate** [comment]
Hi Dave,

I again come across this indicator and the ATR trailing Stops value is not matching.
This time I am calculating ATR on the Heiken Ashi candle and there is a huge difference.
Attaching some files for your reference

Here I am using period 7, multiplier 3

1. Heiken Ashi Candle Data 
[HeikinAshiResult 5minute.csv](https://github.com/DaveSkender/Stock.Indicators/files/9402430/HeikinAshiResult.5minute.csv)

3. Volatility Stop by Skender DLL
[VolatilityStopResult 5minute.csv](https://github.com/DaveSkender/Stock.Indicators/files/9402435/VolatilityStopResult.5minute.csv)

5. Actual value as per the charting platform is 

![image](https://user-images.githubusercontent.com/76814842/186153892-ecafb9a2-a34f-411e-9c72-4852a1defeba.png)

![image](https://user-images.githubusercontent.com/76814842/186153616-4a51c69b-8769-4857-bfbe-23911c332260.png)

Reference for ATR Trailing Stop Signals
https://www.incrediblecharts.com/indicators/atr_average_true_range_trailing_stops.php#:~:text=ATR%20Trailing%20Stops%20Formula,stop%20for%20the%20following%20day

[ATR Trailing Stops Formula](https://www.incrediblecharts.com/indicators/atr_average_true_range_trailing_stops.php#atr-formula)
Trailing stops are normally calculated relative to the closing price:

Calculate [Average True Range](https://www.incrediblecharts.com/indicators/average_true_range.php) ("ATR")
Multiply ATR by your selected multiple — in our case 3 x ATR
In an up-trend, subtract 3 x ATR from Closing Price and plo

t the result as the stop for the following day
If price closes below the ATR stop, add 3 x ATR to Closing Price — to track a Short trade
Otherwise, continue subtracting 3 x ATR for each subsequent day until price reverses below the ATR stop
We have also built in a ratchet mechanism so that ATR stops cannot move lower during a Long trade nor rise during a Short trade.
The HighLow option is a little different: 3xATR is subtracted from the daily High during an up-trend and added to the daily Low during a down-trend.

Reference for Volatility Stops
https://www.incrediblecharts.com/indicators/volatility_stops.php

[Volatility Stops Formula](https://www.incrediblecharts.com/indicators/volatility_stops.php#volatility_formula)
Welles Wilder's system uses Closing Price and incorporates a stop-and-reverse feature (as with his [Parabolic SAR](https://www.incrediblecharts.com/technical/parabolic_sar.htm)).

Determine the initial trend direction
Calculate the Significant Close ("SIC"): the highest close reached in an up-trend or the lowest close in a down-trend
Calculate [Average True Range](https://www.incrediblecharts.com/technical/average_true_range.htm) ("ATR") for the selected period (7 days in this example)
Multiply ATR by the Multiple (3.0 in this example)
The first stop is calculated in day 7 and plotted for day 8
If an up-trend, the first stop is SIC - 3 * ATR, otherwise SIC + 3 * ATR for a down-trend
Repeat each day until price closes below the stop (or above in a down-trend)
Set SIC equal to the latest Close, reverse the trend and continue.

**Reply by DaveSkender** [reply]
I’ll keep this in mind when implementing ATR Trailing Stop in the library, thanks for the extra info.  Helpful.

**Reply by DaveSkender** [reply]
To follow work on this one, see #729 
Also note known discrepancy with other sites in #820 for Volatility Stop.
I did also find some key differences from Volatility Stop in my research as well.

**Reply by myalgomate** [reply]
Hi Dave,

Added the SDK reference of where I am tracking the chart and gave the above example.

https://documentation.chartiq.com/tutorial-Using%20and%20Customizing%20Studies%20-%20Definitions.html#atr-trailing

Average True Range (ATR) Trailing Stop
Indicator Type
Trade Trigger

Formula
Calculate the Average True Range and then multiply by the multiplier. For an up trend, subtract from previous close. For a down trend, add to the previous close.

Math
TR=True Range = defined as the greatest of the following:

Current high minus the current low
Current high minus the previous close (absolute value)
Current low minus the previous close (absolute value)
ATR = simple moving average of TR
ATR Trailing Stop in a rising trend is the highest price over the user-defined span minus (multiplier * ATR). If prices continue to rise, the stop will continue to rise. However, if prices flatten out or pull back the stop will be flat. Only when the price dips below the flat stop line will it flip to the inverse with the stop above prices.

![image](https://user-images.githubusercontent.com/76814842/186347002-fc7569d8-7df2-4f38-b6ec-6e9488fe8873.png)

**Reply by DaveSkender** [reply]
Thanks for the info.  [True Range](https://daveskender.github.io/Stock.Indicators/indicators/Atr/#content) and [Average True Range](https://daveskender.github.io/Stock.Indicators/indicators/Atr/#content) are already in our library, if helpful.

**Reply by myalgomate** [reply]
Yes its added but i really follow your standards.
i was using Wilder's [Volatility Stop](https://daveskender.github.io/Stock.Indicators/indicators/VolatilityStop/#content)
But its not matching the value with ATR trail stops.

It would really help a lot if u add this indicator.

**Reply by DaveSkender** [reply]
Yah, Volatility Stop is different than ATR Trailing Stop, so I can see your need on this one.

**Reply by wc-code-challenges** [reply]
Thanks for adding this one in version 2.3.0 of the library Dave

**Comment by UltimateLaForsch** [comment]
Hello Dave,
a more general question, if I may. In the docs you show some candlestick charts with the indicators. What would be the libraray for displaying the charts, please?
I searched quite a while for this but could not find it...
Thank you
Best Regards
Patrick

**Reply by DaveSkender** [reply]
Here's a couple helpful items:

- [Our demo site](https://charts.stockindicators.dev/) is where I generate those charts.  It's good for seeing what indicators might look like with different parameters.  I'm not necessarily recommending it, but I'm using Angular with Charts.js ([source code](https://github.com/facioquo/stock-charts))
- Our discussion #430 has many more options to consider, if you're interested in integrating or building something yourself.  Plotly seems to be a popular choice recently.

**Reply by UltimateLaForsch** [reply]
Hello Dave,
many thx for the quick answer. Ok, that explains it, it's a frontend solution...
Matplotlib does not have these kind of charts, I am aware of plotly, but I think it does not look as good as yours... maybe PyQt could be worth another look for a pure python solution...
In any case thx again 
P.S.: and maybe I have look at your frontend solution link
